File: packages/sdk-ts/src/resources/appointments.ts
import type { Http, RequestOptions } from '../core/http.js';
import type { Page } from '../core/page.js';
import type { CreateAppointment, Appointment, PaginationQuery, UpdateAppointment } from '../types.js';

const enc = encodeURIComponent;

/** Appointments booked by agents or created directly. */
export class Appointments {
    constructor(private readonly http: Http) {}

    list(query: PaginationQuery = {}, opts?: RequestOptions): Promise<Page<Appointment>> {
        return this.http.list<Appointment>('/v1/appointments', query, opts);
    }

    create(body: CreateAppointment, opts?: RequestOptions): Promise<Appointment> {
        return this.http.request<Appointment>('POST', '/v1/appointments', { ...opts, body });
    }

    retrieve(id: string, opts?: RequestOptions): Promise<Appointment> {
        return this.http.request<Appointment>('GET', `/v1/appointments/${enc(id)}`, opts);
    }

    /** Only the fields sent are changed. */
    update(id: string, body: UpdateAppointment, opts?: RequestOptions): Promise<Appointment> {
        return this.http.request<Appointment>('PATCH', `/v1/appointments/${enc(id)}`, { ...opts, body });
    }

    delete(id: string, opts?: RequestOptions): Promise<void> {
        return this.http.request<void>('DELETE', `/v1/appointments/${enc(id)}`, opts);
    }
}

File: packages/sdk-dart/lib/src/resources/appointments.dart
import '../models/appointment.dart';
import '../page.dart';
import '../transport.dart';

/// Appointments booked by agents or created directly.
class Appointments {
  /// Binds the resource to a transport.
  const Appointments(this._transport);

  final Transport _transport;

  /// Lists appointments.
  Future<Page<Appointment>> list(
          {int? limit, String? startingAfter, String? endingBefore}) =>
      _transport.list('/v1/appointments', Appointment.fromJson, query: {
        'limit': limit,
        'starting_after': startingAfter,
        'ending_before': endingBefore,
      });

  /// Creates one.
  Future<Appointment> create(CreateAppointment body) =>
      _transport.request('POST', '/v1/appointments', Appointment.fromJson,
          body: body.toJson());

  /// Fetches one.
  Future<Appointment> retrieve(String id) => _transport.request('GET',
      '/v1/appointments/${Uri.encodeComponent(id)}', Appointment.fromJson);

  /// Changes one. Only the fields you set are changed.
  Future<Appointment> update(String id, UpdateAppointment body) =>
      _transport.request('PATCH', '/v1/appointments/${Uri.encodeComponent(id)}',
          Appointment.fromJson,
          body: body.toJson());

  /// Deletes one.
  Future<void> delete(String id) => _transport.requestVoid(
      'DELETE', '/v1/appointments/${Uri.encodeComponent(id)}');
}
